From f4f119946a4070f5aa9826c8c41e435760cd81f7 Mon Sep 17 00:00:00 2001
From: Gerald Combs <gerald@wireshark.org>
Date: Fri, 2 Jan 2015 14:06:42 -0800
Subject: [PATCH 4/8] Make sure we don't underrun a buffer when decrypting SSL.

Discovered by Noam Rathaus.

Change-Id: Ia0275601b2a825ba616656064d9a6eca109e34fa
Reviewed-on: https://code.wireshark.org/review/6256
Petri-Dish: Gerald Combs <gerald@wireshark.org>
Reviewed-by: Gerald Combs <gerald@wireshark.org>
Tested-by: Petri Dish Buildbot <buildbot-no-reply@wireshark.org>
Reviewed-by: Evan Huus <eapache@gmail.com>
(cherry picked from commit d3581aecda62d2a51ea7088fd46975415b03ec57)
Reviewed-on: https://code.wireshark.org/review/6324
(cherry picked from commit f5e435ab8bccaf64ca93dc6e2330090b2e99e532)
Reviewed-on: https://code.wireshark.org/review/6428
Reviewed-by: Michael Mann <mmann78@netscape.net>
---
 epan/dissectors/packet-ssl-utils.c | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/epan/dissectors/packet-ssl-utils.c b/epan/dissectors/packet-ssl-utils.c
index 194efea..eb9cb1a 100644
--- a/epan/dissectors/packet-ssl-utils.c
+++ b/epan/dissectors/packet-ssl-utils.c
@@ -3193,6 +3193,10 @@ ssl_decrypt_record(SslDecryptSession*ssl,SslDecoder* decoder, gint ct,
 
     /* strip padding for GenericBlockCipher */
     if (decoder->cipher_suite->mode == MODE_CBC) {
+        if (inl < 1) { /* Should this check happen earlier? */
+            ssl_debug_printf("ssl_decrypt_record failed: input length %d too small\n", inl);
+            return -1;
+        }
         pad=out_str->data[inl-1];
         if (worklen <= pad) {
             ssl_debug_printf("ssl_decrypt_record failed: padding %d too large for work %d\n",
-- 
2.1.4

